.github/workflows/reusable_terraform_plan_apply_test.yml

```diff
@@ -0,0 +1,287 @@
+        working-directory: "terraform/environments/${{ inputs.application }}"
+        run: |
+          set -o pipefail
+          exitcode=0
+          tf_args="-detailed-exitcode ${{ inputs.init_plan_apply_tfargs }} ${{ inputs.plan_apply_tfargs }}"
+          [[ ${POST_PLAN_TO_PR} == 'true' ]] && tf_args="${tf_args} -no-color"
+          [[ ${{ inputs.do_state_refresh_on_plan }} == 'true' ]] && tf_args="${tf_args} -refresh=false"
+          echo "terraform plan ${tf_args}"
+          terraform plan ${tf_args} | bash ${GITHUB_WORKSPACE}/scripts/redact-output.sh | tee tfplan.txt || exitcode=$?
+          echo "exitcode=${exitcode}"  # 0=clean plan, 1=error, 2=stuff in plan
+          echo "exitcode=${exitcode}" >> $GITHUB_OUTPUT
+          (( exitcode == 1 )) && exit 1 || exit 0
+      - name: Create Plan PR message (Optional)
+        if: github.event_name == 'pull_request' && steps.plan.outputs.exitcode == '2' && inputs.post_plan_to_pr == true
+        working-directory: "terraform/environments/${{ inputs.application }}"
+        run: |
+          comment() {
+            url="https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}"
+            len=$(cat tfplan.txt | wc -c)
+            echo '**`${{ env.WORKSPACE_NAME }}`** terraform plan on `${{ github.event_name }}` event [#${{ github.run_number }}]('${url}')'
+            echo
+            echo '```'
+            head -c 65476 tfplan.txt | sed -n '/Terraform will perform/,$p'
+            echo
+            echo '```'
+            if [[ $len -gt 65476 ]]; then
+              echo "** Truncated output.  See $url for the rest **"
+            fi
+          }
+          echo 'TF_PLAN_OUT<<EOF' >> $GITHUB_ENV
+          comment >> $GITHUB_ENV
+          echo 'EOF' >> $GITHUB_ENV
+      - name: Hide Previous PR comment (Optional)
+        if: ${{ github.event_name == 'pull_request' }}
+        working-directory: "scripts/minimise-comments"
+        env:
+          COMMENT_BODY_CONTAINS: "**`${{ env.WORKSPACE_NAME }}`**"
+          PR_NUMBER: "${{ github.event.pull_request.number }}"
+        run: |
+          go build
+          ./minimise-comments
+      - name: Post Plan to PR (Optional)
+        if: github.event_name == 'pull_request' && steps.plan.outputs.exitcode == '2' && inputs.post_plan_to_pr == true
+        env:
+          message: "${{ env.TF_PLAN_OUT }}"
+        run: |
+          escaped_message=$(echo "$message" | jq -Rsa .)
+          curl -sS -X POST \
+            -H "Accept: application/vnd.github+json" \
+            -H "Authorization: Bearer ${{ env.GITHUB_TOKEN }}" \
+            "https://api.github.com/repos/${{ github.repository }}/issues/${{ github.event.pull_request.number }}/comments" \
+            -d '{"body":'"${escaped_message}"'}'
+  terratest:
+    name: "terratest"
+    needs: plan
+    if: inputs.action == 'plan_apply' && needs.plan.outputs.plan_exitcode == '2' && inputs.environment == 'development'
```

Comment:
is exit code 2 "succeeded with non-empty diff" here? https://developer.hashicorp.com/terraform/cli/commands/plan#detailed-exitcode

If so I think we should also check for code 0 as well

Reply:
from the terraform plan step:
`          echo "exitcode=${exitcode}"  # 0=clean plan, 1=error, 2=stuff in plan`

would we want to run terraform tests when there's a clean plan?

Reply:
Hmm maybe not the terratest ones, but the smoke tests still seem relevant if you have just deployed something. What do you think?

Reply:
I'd run the smoke tests after each successful apply?